pkg/graveler/graveler.go

```diff
@@ -2528,6 +2530,124 @@ func (g *Graveler) Merge(ctx context.Context, repository *RepositoryRecord, dest
 	return commitID, nil
 }
 
+func (g *Graveler) MergeFromMetaRange(ctx context.Context, repository *RepositoryRecord, destination BranchID, source MetaRangeID, commitParams CommitParams) (CommitID, error) {
+	var (
+		preRunID string
+		commit   Commit
+		commitID CommitID
+	)
+
+	storageNamespace := repository.StorageNamespace
+	if err := g.prepareForCommitIDUpdate(ctx, repository, destination); err != nil {
+		return "", err
+	}
+
+	var tokensToDrop []StagingToken
+	// No retries on any failure during the merge. If the branch changed, it's either that commit is in progress, commit occurred,
+	// or some other branch changing operation. If commit is in-progress, then staging area wasn't empty after we checked so not retrying is ok.
+	// If another commit/merge succeeded, then the user should decide whether to retry the merge.
+	err := g.retryBranchUpdate(ctx, repository, destination, func(branch *Branch) (*Branch, error) {
+		empty, err := g.isSealedEmpty(ctx, repository, branch)
+		if err != nil {
+			return nil, fmt.Errorf("check if staging empty: %w", err)
+		}
+		if !empty {
+			return nil, fmt.Errorf("%s: %w", destination, ErrDirtyBranch)
+		}
+		toCommit, err := g.dereferenceCommit(ctx, repository, Ref(destination))
+		if err != nil {
+			return nil, err
+		}
+
+		// Source is not a commit - merge base is destination root
+		itr, err := g.Log(ctx, repository, toCommit.CommitID, false)
+		if err != nil {
+			return nil, err
+		}
+		root := toCommit
+		for itr.Next() {
+			root = itr.Value()
+		}
+
+		g.log(ctx).WithFields(logging.Fields{
+			"repository":             repository.RepositoryID,
+			"source":                 source,
+			"destination":            destination,
+			"source_meta_range":      source,
+			"destination_meta_range": toCommit.MetaRangeID,
+			"base_meta_range":        root,
+		}).Trace("Merge")
+
+		metaRangeID, err := g.CommittedManager.MergeFromMetaRange(ctx, storageNamespace, toCommit.MetaRangeID, root.MetaRangeID, source, MergeStrategySrc)
+		if err != nil {
+			if !errors.Is(err, ErrUserVisible) {
+				err = fmt.Errorf("merge in CommitManager: %w", err)
+			}
+			return nil, err
+		}
+		commit = NewCommit()
+		commit.Committer = commitParams.Committer
+		commit.Message = commitParams.Message
+		commit.MetaRangeID = metaRangeID
+		commit.Parents = []CommitID{toCommit.CommitID}
+		commit.Generation = toCommit.Generation + 1
+		commit.Metadata = commitParams.Metadata
+		commit.Metadata[MergeStrategyMetadataKey] = MergeStrategySrcWinsStr
+		preRunID = g.hooks.NewRunID()
+		err = g.hooks.PreMergeHook(ctx, HookRecord{
```

Comment:
we trigger merge event with commit information that holds single parent

Reply:
That's the result of merging on a metarange instead of a commit...

Reply:
The user perform import and see a merge commit with merge hook. If we are not keeping the side branch why do we keep communicate an merge event instead of a commit?